content/intro-to-storybook/ember/en/composite-component.md

```diff
@@ -0,0 +1,301 @@
+
+![Qunit logo](/intro-to-storybook/logo-qunit.jpg)
+
+### Unit tests with Qunit
+
+Storybook stories paired with manual visual tests go a long way to avoiding UI bugs. If stories cover a wide variety of component use cases, and we use tools that ensure a human checks any change to the story, errors are much less likely.
+
+However, sometimes the devil is in the details. A test framework that is explicit about those details is needed. Which brings us to unit tests.
+
+In our case, we want our `TaskList` to render any pinned tasks **before** unpinned tasks that it has passed in the `tasks` prop. Although we have a story (`WithPinnedTasks`) to test this exact scenario, it can be ambiguous to a human reviewer that if the component **stops** ordering the tasks like this, it is a bug. It certainly won’t scream **“Wrong!”** to the casual eye.
+
+So, to avoid this problem, we can use Qunit to render the component and run some DOM querying code to verify salient features of the output.
+
+Create a test file called `tests/integration/task-list-test.js`. Here, we’ll build out our tests that make assertions about the output.
+
+```javascript
+// tests/integration/task-list-test.js
+import { module, test } from 'qunit';
+import { setupRenderingTest } from 'ember-qunit';
+import { render } from '@ember/test-helpers';
+import { hbs } from 'ember-cli-htmlbars';
+module('Integration | Component | TaskList', function(hooks) {
+  setupRenderingTest(hooks);
+  const taskData = {
+    id: '1',
+    title: 'Test Task',
+    state: 'TASK_INBOX',
+    updatedAt: new Date(2018, 0, 1, 9, 0),
+  };
+  const tasklist = [
+    { ...taskData, id: '1', title: 'Task 1' },
+    { ...taskData, id: '2', title: 'Task 2' },
+    { ...taskData, id: '3', title: 'Task 3' },
+    { ...taskData, id: '4', title: 'Task 4' },
+    { ...taskData, id: '5', title: 'Task 5' },
+    { ...taskData, id: '6', title: 'Task 6 (pinned)', state: 'TASK_PINNED' },
+  ];
+
+  test('renders pinned tasks at the start of the list', async function(assert) {
+    this.set('tasks', tasklist);
+    this.set('testpinTask', actual => {
+      let expected = 1;
+      assert.deepEquals(actual, expected);
+    });
+
+    this.set('testarchiveTask', actual => {
+      let expected = 1;
+      assert.deepEquals(actual, expected);
+    });
+
+    await render(
+      hbs`{{TaskList tasks=tasks pinTask=(action testpinTask) archiveTask=(action testarchiveTask) }}`
+    );
+    assert.dom(this.element.querySelector('.list-item:nth-of-type(1)')).hasClass('TASK_PINNED');
+  });
+});
+```
+
+![TaskList test runner](/intro-to-storybook/tasklist-testrunner-ember.png)
+
+Contrary to the other versions of this tutorial, with Ember we can't import the data and stories used in our story file created earlier, without introducing a lot of complexity and that's beyond the scope of the tutorial. For now we'll copy over the values used in the story file to help out with our tests.
```

Comment:
Out of interest what is the challenge in importing + using the stories? Maybe this is something we should fix in Storybook?

Reply:
@tmeasday upon a bit of "detective work" (pardon the bad pun) it seems that jest is not working with ember as of present state of frameworks. So having to test it in Qunit was something that was not that trivial as i expected. A bit more context here. Normally with jest you import what you need into the test file and use it's contents at your discretion, like it's documented in other versions of the tutorial correct? With ember/Qunit, i tried the same approach and in vscode the file was being resolved without a hitch. I ran the test command and it was not being picked up by the test runner, it wouldn't resolve. I tinkered a bit to see what how i could use it. But i kept having a feeling that i was forgetting about something. And i was correct, back in the day when i used ember to "mock" external files/date you'll need to copy over the file into `vendor` and configure `ember-cli-build.js` to use the file. 
Something like so:
```js
'use strict';

const EmberApp = require('ember-cli/lib/broccoli/ember-app');

module.exports = function(defaults) {
  let app = new EmberApp(defaults, {
    // Add options here
  });

  /* if (app.env==='test'){
    app.import('vendor/taskbox-fixtures/Task.stories.js',{
      type: 'vendor',
      prepend: true
    });
    app.import('vendor/taskbox-fixtures/TaskList.stories.js',{
      type: 'vendor',
      prepend: true
    });
  } */
  

  // Use `app.import` to add additional libraries to the generated
  // output files.
  //
  // If you need to use different assets in different
  // environments, specify an object as the first parameter. That
  // object's keys should be the environment name and the values
  // should be the asset to use in that environment.
  //
  // If the library that you are including contains AMD or ES6
  // modules that you would like to import into your application
  // please specify an object with the list of modules as keys
  // along with the exports of each module as its value.

  return app.toTree();
};
```
(ignore the comments this was taken out of the code used to back the documentation)

When me or the reader issues the test command the file will be resolved, but not Storybook and/or Webpack. I tried a couple of ways to have the modules added here so i could use the file but it wasn't that trivial and also i wasn't able to get it running with success. Hence why i went with that phrase to provide a bit of context and allow the reader to have a simple test added.

Reply:
Hmm, I see. I guess the idea of that section is to show how to do unit tests *using the stories* we've created. 

Although it is useful to know how to write a generic unit test for UI code, I'm not sure if it's something that a tutorial for Storybook should focus on. I suspect we might want to show people how to get their CSF files included in stories, even if it is complex. WDYT?